File: Roadmap/07 - PILAS Y COLAS/python/juanppdev.py
"""
Implementa los mecanismos de introducción y recuperación de elementos propios de las
 * pilas (stacks - LIFO) y las colas (queue - FIFO) utilizando una estructura de array
 * o lista (dependiendo de las posibilidades de tu lenguaje).
"""

"""
Stack LIFO
"""

class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()

    def peek(self):
        if not self.is_empty():
            return self.items[-1]

    def is_empty(self):
        return len(self.items) == 0

    def size(self):
        return len(self.items)
    
stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)
print(stack.peek()) # prints: 3
print(stack.pop())  # prints: 3
print(stack.size()) # prints: 2

"""
Queue (FIFO)
"""

class Queue:
    def __init__(self):
        self.items = []

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)

    def peek(self):
        if not self.is_empty():
            return self.items[0]

    def is_empty(self):
        return len(self.items) == 0

    def size(self):
        return len(self.items)
    
queue = Queue()
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
print(queue.peek()) # prints: 1
print(queue.dequeue()) # prints: 1
print(queue.size()) # prints: 2

"""
EXTRA
"""

class NavegadorWeb:
    def __init__(self):
        self.pila = []
        self.actual = None

    def navegar(self, sitio):
        if self.actual is not None:
            self.pila.append(self.actual)
        self.actual = sitio
        print("Navegando a:", sitio)

    def adelante(self):
        if self.pila:
            sitio = self.pila.pop()
            print("Avanzando hacia adelante a:", sitio)
            self.actual = sitio
        else:
            print("No hay páginas para avanzar")

    def atras(self):
        if self.actual is not None:
            print("Retrocediendo hacia atrás a:", self.actual)
            self.actual = None
        else:
            print("No hay página actual")

# Probando la implementación
navegador = NavegadorWeb()
navegador.navegar("www.ejemplo1.com")
navegador.navegar("www.ejemplo2.com")
navegador.adelante()
navegador.atras()
navegador.adelante()


from collections import deque

class ImpresoraCompartida:
    def __init__(self):
        self.cola_impresion = deque()

    def agregar_documento(self, documento):
        self.cola_impresion.append(documento)
        print("Documento agregado a la cola:", documento)

    def imprimir(self):
        if self.cola_impresion:
            documento = self.cola_impresion.popleft()
            print("Imprimiendo documento:", documento)
        else:
            print("No hay documentos para imprimir")

# Probando la implementación
impresora = ImpresoraCompartida()
impresora.agregar_documento("Documento1")
impresora.agregar_documento("Documento2")
impresora.imprimir()
impresora.imprimir()

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/07 - PILAS Y COLAS/swift/juanppdev.swift
/*
    Implementa los mecanismos de introducción y recuperación de elementos propios de las
 * pilas (stacks - LIFO) y las colas (queue - FIFO) utilizando una estructura de array
 * o lista (dependiendo de las posibilidades de tu lenguaje).
*/
// Implementación de una pila (stack) utilizando un array
struct Stack<T> {
    private var elements = [T]()
    
    mutating func push(_ element: T) {
        elements.append(element)
    }
    
    mutating func pop() -> T? {
        return elements.popLast()
    }
    
    func peek() -> T? {
        return elements.last
    }
    
    func isEmpty() -> Bool {
        return elements.isEmpty
    }
}

// Implementación de una cola (queue) utilizando un array
struct Queue<T> {
    private var elements = [T]()
    
    mutating func enqueue(_ element: T) {
        elements.append(element)
    }
    
    mutating func dequeue() -> T? {
        if elements.isEmpty {
            return nil
        } else {
            return elements.removeFirst()
        }
    }
    
    func peek() -> T? {
        return elements.first
    }
    
    func isEmpty() -> Bool {
        return elements.isEmpty
    }
}

// Ejemplo de uso de la pila
var stack = Stack<Int>()
stack.push(1)
stack.push(2)
stack.push(3)

while !stack.isEmpty() {
    if let element = stack.pop() {
        print("Elemento extraído de la pila: \(element)")
    }
}

// Ejemplo de uso de la cola
var queue = Queue<String>()
queue.enqueue("Hola")
queue.enqueue("Mundo")

while !queue.isEmpty() {
    if let element = queue.dequeue() {
        print("Elemento extraído de la cola: \(element)")
    }
}



// Extra

// Simulación de un navegador web utilizando una pila
func simulateWebBrowser() {
    var history = Stack<String>()
    
    func navigate(to page: String) {
        print("Navegando a la página: \(page)")
        history.push(page)
    }
    
    func goBack() {
        if let previousPage = history.pop() {
            print("Volviendo a la página anterior: \(previousPage)")
        } else {
            print("No hay páginas anteriores en el historial.")
        }
    }
    
    // Simulación de la navegación
    navigate(to: "Google")
    navigate(to: "Facebook")
    navigate(to: "Twitter")
    
    goBack()
    goBack()
}

// Simulación de una impresora compartida utilizando una cola
func simulateSharedPrinter() {
    var printerQueue = Queue<String>()
    
    func printDocument(_ document: String) {
        print("Imprimiendo documento: \(document)")
    }
    
    func processPrintQueue() {
        if let document = printerQueue.dequeue() {
            printDocument(document)
        } else {
            print("No hay documentos en la cola de impresión.")
        }
    }
    
    // Simulación de la cola de impresión
    printerQueue.enqueue("Documento1.pdf")
    printerQueue.enqueue("Documento2.docx")
    printerQueue.enqueue("Documento3.txt")
    
    processPrintQueue()
    processPrintQueue()
}

// Ejecutar simulaciones
simulateWebBrowser()
simulateSharedPrinter()